problem-gson/src/test/java/org/zalando/problem/GsonProblemTest.java

```diff
@@ -0,0 +1,306 @@
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.net.URI;
+import java.net.URL;
+import java.util.List;
+import java.util.Objects;
+
+import static com.jayway.jsonassert.JsonAssert.with;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasEntry;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.hasToString;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+import static org.hamcrest.Matchers.startsWith;
+import static org.hobsoft.hamcrest.compose.ComposeMatchers.hasFeature;
+import static org.zalando.problem.Status.BAD_REQUEST;
+
+class GsonProblemTest {
+
+    private static Gson gson = new GsonBuilder()
+            .registerTypeAdapterFactory(new ProblemAdapterFactory()
+                    .registerSubtype(InsufficientFundsProblem.class, InsufficientFundsProblem.TYPE)
+                    .registerSubtype(OutOfStockException.class, OutOfStockException.TYPE))
+            .create();
+
+    private static JsonReader getReader(final String name) throws FileNotFoundException {
+        final ClassLoader loader = Thread.currentThread().getContextClassLoader();
+        URL resource = Objects.requireNonNull(loader.getResource(name), () -> "resource " + name + " not found.");
+        return new JsonReader(new FileReader(new File(resource.getPath())));
+    }
+
+    private String getStackTrace(final Throwable throwable) {
+        final StringWriter writer = new StringWriter();
+        throwable.printStackTrace(new PrintWriter(writer));
+        return writer.toString();
+    }
+
+    @Test
+    void shouldSerializeDefaultProblem() {
+        final Problem problem = Problem.valueOf(Status.NOT_FOUND);
+        final String json = gson.toJson(problem);
+        assertThat("json", Objects.nonNull(json));
+        with(json)
+                .assertThat("$.*", hasSize(2))
+                .assertThat("$.title", is("Not Found"))
+                .assertThat("$.status", is(404));
+    }
+
+    @Test
+    void shouldSerializeCustomProperties() {
+        final Problem problem = Problem.builder()
+                .withType(URI.create("https://example.org/out-of-stock"))
+                .withTitle("Out of Stock")
+                .withStatus(BAD_REQUEST)
+                .withDetail("Item B00027Y5QG is no longer available")
+                .with("product", "B00027Y5QG")
+                .build();
+
+        final String json = gson.toJson(problem);
+        assertThat("json", Objects.nonNull(json));
+        with(json)
+                .assertThat("$.*", hasSize(5))
+                .assertThat("$.product", is("B00027Y5QG"));
+    }
+
+    @Test
+    void shouldSerializeCustomProblem() {
```

Comment:
Is it possible to serialize a custom problem that has not been registered before?

Reply:
Yes, serialization process doesn't really refers to registered types. With exception of `DefaultProblem`, everything else is delegated to Gson.

Reply:
Can you add a test case for that?

Reply:
Sure, will do that 👍